### DISCUSSION TITLE
Yarn Install result in code server timed out

### DISCUSSION THREAD

**Discussion by 7HE-W0R1D** [coder/code-server#2692] [Q&A] [answered]
I installed code server and an unofficial node 15 on my raspberry pi zero. It can start normally and I can access it with https over browser. However, several dependencies and modules were missing and when I was installing "spdlog", I accidentally installed the official build of node, which is only node 10 on my armv6. Then I reinstalled node 15 but whenever I type code-server, it responses with a `error timed out` I also tried deleting the code-server folders and run `yarn add code-server` again but the error persists. I have no idea about how to fix that since there are no extra info provided by the code server. Any help would be appreciated!

**Comment by code-asher** [comment, marked as answer]
Weird, that's the only thing it outputs?

**Reply by 7HE-W0R1D** [reply]
Sadly yes, it shows this every time
![image](https://user-images.githubusercontent.com/42311695/107480096-0579fd00-6bb7-11eb-8922-c4603fcc38b1.png)

**Reply by code-asher** [reply]
Wow talk about an unhelpful error. It doesn't seem to be getting far enough to even start printing logs which is weird.

Do flags like `code-server --help` or `code-server --version` work or does those also error?

**Reply by code-asher** [reply]
Ah I think the error is coming from the handshake with the child process that code-server spawns. 

I have no idea why the handshake times out though. Could you run `code-server --log trace` and see if anything else gets output?

**Reply by 7HE-W0R1D** [reply]
tried those and yes `code-server --help` or `code-server --version` do work properly, but `code-server --log trace` still produces the same error.

**Reply by 7HE-W0R1D** [reply]
I'm suspecting that I haven't uninstalled the code server fully at the very beginning. Cannot remember clearly but I think I just deleted  node_modules and package.json and ran `yarn add code-server` at the same location. I will try to look into that.

**Reply by 7HE-W0R1D** [reply]
Ah still no luck getting it working, guess I'll do a clean install of the system and then install code server :(

**Reply by 7HE-W0R1D** [reply]
I just did a clean install of the system, nodejs and basically everything. Now code-server can start the server successfully but I'm getting the error " exited unexpectedly with code 0" now, even "code-server --log trace" produces no extra error messages. 
![image](https://user-images.githubusercontent.com/42311695/107875260-d2f33b80-6ef9-11eb-8ec8-867064dace8e.png)

**Reply by 7HE-W0R1D** [reply]
oh noticed that there are only node_modules in the folder I installed code-server not the two (or three) like the last time. I guess there are lots of things missing. Trying to do a reinstall

**Reply by 7HE-W0R1D** [reply]
nope reinstall won't work, still getting the same error.

**Reply by code-asher** [reply]
Hmm there could be some more logs from VS Code in `~/.local/share/code-server/logs/[latest date]` but it might be a long shot.

Are there Node modules in the VS Code directory? Should be `[code server install directory]/lib/vscode/node_modules`. If there are running `npm rebuild` in the `lib/vscode` directory might help but that's another long shot.

**Reply by 7HE-W0R1D** [reply]
> Hmm there could be some more logs from VS Code in `~/.local/share/code-server/logs/[latest date]` but it might be a long shot.
> 
> Are there Node modules in the VS Code directory? Should be `[code server install directory]/lib/vscode/node_modules`. If there are running `npm rebuild` in the `lib/vscode` directory might help but that's another long shot.

Oops sorry forget to mention there, I changed to a more stable npm mirror and the node modules(at least lots of them, cannot find a place showing all the required modules) were back. For the 500 error I think its related the the latest 3.9.0 because even after I get most of the modules, I still get the error with no extra messages, but if I run `yarn global add code-server@3.8.0` then I get the error messages I want. I've opened an issue here https://github.com/cdr/code-server/issues/2732#issue-809336045
Anyway this great server's 3.8.0 version is working pretty well on my little pi zero, the two thing left is that I cannot find a way to set up git. It says "no souce control providers regiestered" even after I ran` apt-get install git npm install git` and installed the git extension pack. Any help is appreciated!

**Reply by 7HE-W0R1D** [reply]
Also forgive me for another silly question, I cannot get my codes (java & python) run in the server. I have python 3, openjdk 8, python & java extension packs installed. In the webpage I can see a run button for the java code, but after I click it nothing will happen. For python files there are no run button at all. What should I do to fix that?

**Reply by code-asher** [reply]
Ah yeah we have an issue where it keeps going even if the dependencies fail to install: https://github.com/cdr/code-server/issues/2651

That's all you should have to do for git. Can double-check if VS Code can find git? You can do something like open a terminal in VS Code and run `which git`.

For Python and Java I recommend you download and install manually from OpenVSX (https://open-vsx.org/). We currently have some issues with our marketplace where the extensions we provide are broken. Soon we'll switch to OpenVSX by default.

**Reply by 7HE-W0R1D** [reply]
> Ah yeah we have an issue where it keeps going even if the dependencies fail to install: https://github.com/cdr/code-server/issues/2651
> 
> That's all you should have to do for git. Can double-check if VS Code can find git? You can do something like open a terminal in VS Code and run `which git`.
> 
> For Python and Java I recommend you download and install manually from OpenVSX (https://open-vsx.org/). We currently have some issues with our marketplace where the extensions we provide are broken. Soon we'll switch to OpenVSX by default.

wow @code-asher thanks for the reply! I tried but the git solution brings another issue out -- I cannot open terminal in vscode web. I can switch to the terminal tab but nothing will display there. I googled for it and find out it has somethings to do with the settings.json. However my setting.json is always empty, even after delete and creating a default one. Is that normal? I think if I can bring up the terminal then everything will be fine since I can get an output if I run `which git` in raspi's terminal.
It's late midnight here so I'll try the vsx extensions tomorrow and update here!

**Reply by code-asher** [reply]
That's definitely odd and shouldn't happen. Are there any logs from code-server's output or in the browser console when you try to open the terminal?

**Reply by 7HE-W0R1D** [reply]
hmm nope, it didn't show anything. I also looked at the logs here at `~/.local/share/code-server/logs/[latest date]` but it seems that the server doesn't record any of my actions when clicking the terminal, creating a new window and terminating a terminal window. However good news is that I managed to get git & github show in the source control panel. Though there are no `publish to github` button, only a `create repository` which when I click does not do anything.

**Reply by code-asher** [reply]
Interesting. If the terminal isn't showing up it could be because the native Node module had problems being built. Can you confirm whether you have all these dependencies? https://github.com/microsoft/vscode-wiki/blob/main/How-to-Contribute.md#prerequisites

If not, installing them and running `npm rebuild` in the `lib/vscode` directory of code-server might resolve the issues.

**Reply by code-asher** [reply]
This link might be helpful as well: https://github.com/cdr/code-server/blob/v3.9.0/docs/npm.md

**Reply by code-asher** [reply]
One more thing, you might need to downgrade Node to 12.

**Reply by 7HE-W0R1D** [reply]
Thanks for that! Will check right away!

**Reply by 7HE-W0R1D** [reply]
It turned out that I did missed some dependencies, installed them , downgraded to node 12.21.0, and did a `npm rebuild --unsafe-perm --build-from-source`, it took npm for some two hours to finish but I still cannot see the terminal showing up. Are there any methods I can use to determine where had gone wrong?

**Reply by code-asher** [reply]
Ah that's unfortunate! Did you restart code-server after the rebuild? I assume there are still no error messages in code-server's stdout?

I'm not sure what else we could try except for attaching a debugger and stepping through the code. Or maybe it's crashing somewhere and a core dump could help. https://docs.faircom.com/doc/knowledgebase/76189.htm

**Reply by 7HE-W0R1D** [reply]
Ah just noticed in stdout that after rebuild, my coder server updated itself to 3.9.0, should I downgrade it to 3.8.0 manually?

**Reply by 7HE-W0R1D** [reply]
Seems that this time there are things appearing in the stdout & stderr files. I'm looking it up and I attached it below in case you might need.
[code-server-stdout.log](https://github.com/cdr/code-server/files/6039482/code-server-stdout.log)
[code-server-stderr.log](https://github.com/cdr/code-server/files/6039483/code-server-stderr.log)

**Reply by code-asher** [reply]
Hmm there seem to be quite a few issues with native Node modules being compiled with the wrong version of Node. Are you sure you ran `npm rebuild` with Node 12.21.0? It says they were compiled with 15.0.0 or above.

**Reply by code-asher** [reply]
It should be safe to keep code-server at 3.9.0.

**Reply by 7HE-W0R1D** [reply]
hmm thats weird, I'm pretty sure I ran the npm rebuild after checked node -version, will it help to run the rebuild once again at the same location?

**Reply by 7HE-W0R1D** [reply]
Hmm, there is a chance that I ran in the wrong location, I closed the terminal so I cannot check for that. Is it fine to install in user pi's folder while run it as root?

**Reply by code-asher** [reply]
Yup a rebuild in the same location should work. For `yarn` you can use `yarn global dir` to double-check where it is (if you used `yarn global add`).

You could also try starting from zero:
1. Uninstall code-server completely (maybe delete cache too).
2. Run `node --verson` to confirm you're on v12.
3. Run `yarn node --version` to confirm yarn is also using v12.
4. Run `yarn global add code-server` again.
5. Make sure the output doesn't have any messages about failing to download dependencies.

**Reply by 7HE-W0R1D** [reply]
wow thanks for the fast reply! I just tried to run in the right location and now it's giving me some normal build output. I noticed that there are couple of warnings about a depreciated package, is that might be related to this question?

**Reply by 7HE-W0R1D** [reply]
seems that there are some modules missing, and npm is giving me some `host/IP does not match certificate name errors, not sure why this happens, maybe nginx is messing things up? Will look into that tomorrow morning (1 a.m. now here).

**Reply by code-asher** [reply]
Sounds good. :smile: 

Deprecated warnings should be safe to ignore.

The certificate warning is worrisome. Sounds like it's causing the dependencies to fail to download. It could be a DNS issue? https://stackoverflow.com/a/52129056

**Reply by 7HE-W0R1D** [reply]
Trying to start from zero with node 12 & code server 3.9.0, hope this time there will be a difference.

**Reply by 7HE-W0R1D** [reply]
Also I'm always getting this error when install & remove coer-server, is it ok to have this dependency skipped? 
`info "fsevents@2.3.2" is an optional dependency and failed compatibility check. Excluding it from installation.`

**Reply by 7HE-W0R1D** [reply]
Ah still no luck getting the terminal working. But at least 3.9.0 now runs :)

**Reply by 7HE-W0R1D** [reply]
super weird that I ran npm rebuild with two different mirrors they both give error messages, however one is socket timeout the other is empty response. And the timeout one will appear on different packages if you run it multiple times. I'll try another mirror tomorrow. But I just wonder does that really relate to the terminal problem? Since it just finished with a success message and nothing really changed after the rebuild.

**Reply by code-asher** [reply]
`fsevents` is for MacOS I think so it should be safe to ignore on a Pi.

The terminal requires some Node modules to run so if the timeouts are related to those modules then it would prevent the terminal from working. :thinking: But it's definitely possible the terminal is failing for a different reason.

When this happens I think yarn is supposed to exit with an error but for some reason right now it just keeps going on happily. We have an issue open to fix this: https://github.com/cdr/code-server/issues/2651

One thing that could help is `yarn --check-files`. For example, assuming you installed using `yarn global add code-server` you could try:

```
$ cd $(yarn global dir)/node_modules/code-server/lib/vscode
$ yarn --check-files
```

If you have all the dependencies it should do nothing, otherwise it will try to install them.

**Reply by 7HE-W0R1D** [reply]
Thanks for that! I'm just manually installing the package with connection errors and it's now building code-server & spdlog, which are exactly those yield errors when running the server. Hope this time there will be a difference ;)

**Reply by 7HE-W0R1D** [reply]
Yay it worked!!!!! After like an afternoon building the package terminal now shows up and I can type things in it!!! Though it's a bit laggy (2-3 sec) but I'm sooooo glad it works! Now there's only extensions to deal with ;)
Edit: the lag is due to my NAT traversal service, when connected locally there are no noticeable lags.

**Reply by 7HE-W0R1D** [reply]
Now that the server is totally usable, github's back, python's able to run through command line(no triangle run button, probably lacking some extensions? Mentioned it [here](https://github.com/cdr/code-server/issues/2341#issuecomment-787104617)), and Java files are running smoothly. My NAT traversal service is a bit unstable so connect through weblink is not so fast and sometimes disconnects, but connecting locally now really worked flawlessly.
Huge thanks to @code-asher for this great project and all the time & effort you spent these days helping me!

**Reply by 4oo4** [reply]
I had this same issue, but for some reason I just had to remove old certificates from `~/.local/share/code-server` to fix it. Is `self-signed.cert` the name that Code Server still uses for creating certificates (the docs seem to say that it's `self-signed.cert`)? I also have a `localhost.cert` in that same folder, the localhost one was still valid whereas the self-signed ones had been expired for a while.

I was able to launch Code Server right after upgrading to 3.9.0, but for some reason the issue didn't appear until about a week afterward. My hunch is that the two sets of certificates was potentially causing issues.

Cheers

**Reply by code-asher** [reply]
@4oo4 We changed the generated certificate and key file names. They are now:

1. ~/.local/share/code-server/$hostname.crt
2. ~/.local/share/code-server/$hostname.key

`$hostname` is taken from `--cert-host` and falls back to `localhost` if not set.

It does seem strange the old files could cause an issue though. We don't load them or anything like that. :thinking:

**Reply by 4oo4** [reply]
@code-asher That makes sense. I'll keep an eye on it and see if I keep getting it, since after first upgrading to 3.9.0 I didn't have any issues like that, and the old certs have been sitting out there for a while. So there's a chance it could have been something else, like me restarting my docker container that actually fixed it. If it recurs I'll open a new issue.

Thanks
